### PR TITLE
mavlink: introduce the mavlink protocol implementation

### PR DESCRIPTION
**Pull request by dorileo**
### Changes

+ since  v1 (#1171):
  - fixed issues pointed on #1171;
  - added new api's;
  - fixed issue with doxygen;

+ since v2 (#1207):
  - implemented serial connection;
  - removed an arbitrary throttle argument on speed set call;
  - fixed style check breakage;
  - upstream rebased;

### Dependencies
+ the ```mission_reached``` callbacks on samples require a patch on APM (sent upstream here https://github.com/diydrones/ardupilot/pull/3360).

### Rationale
This patch introduce the basic infrastructure and API's to control MAV
(Micro Air Vehicle) communication. More API's are to come.

Signed-off-by: Leandro Dorileo <leandro.maciel.dorileo@intel.com>

### CODE REVIEW

File: src/lib/comms/include/sol-mavlink.h [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+ *
+ * @see sol_mavlink_connect
+ *
+ * This object is the abstraction of a Mavlink connection. This is the base
+ * structure for all Mavlink operations and is obtained through the
+ * sol_mavlink_connect API.
+ */
+struct sol_mavlink;
+
+typedef enum {
```
Comment (bdilly) [inline src/lib/comms/include/sol-mavlink.h]: enums should be documented as well

also we "are not" using typedefs... ok, you'll find a few spread over our codebase, but we were avoiding it.

File: src/samples/mavlink/goto.c [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+    sol_run();
+    sol_mavlink_disconnect(mavlink);
+    sol_shutdown();
+    return EXIT_SUCCESS;
+
+err:
+    sol_shutdown();
+    return EXIT_FAILURE;
+}
```
Comment (bdilly) [inline src/samples/mavlink/goto.c]: lot's of code duplication between basic and goto samples

maybe you could have a common implementation?

also a description about these samples could be nice, making clear the difference between these samples

File: src/lib/comms/include/sol-mavlink.h [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+    SOL_MAVLINK_MODE_ACRO = 1,
```
Comment (bdilly) [inline src/lib/comms/include/sol-mavlink.h]: not sure if all of these fields require description, but at least a few of these do, don't you think?

File: src/lib/comms/include/sol-mavlink.h (hunk 1 above)
Comment (dorileo) [inline src/lib/comms/include/sol-mavlink.h]: ok

File: src/samples/mavlink/goto.c (hunk 2 above)
Comment (dorileo) [inline src/samples/mavlink/goto.c]: @bdilly I wanted to keep things self contained and easier to follow - to be used as documentation for example.

I'll add a more verbose description to it.

File: src/lib/comms/include/sol-mavlink.h (hunk 3 above)
Comment (dorileo) [inline src/lib/comms/include/sol-mavlink.h]: yep, sure.

File: src/samples/mavlink/goto.c [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    if (err < 0) {
+        SOL_ERR("Could not get current position: %s", sol_util_strerrora(-err));
+        return;
+    }
+
+    if (mode != SOL_MAVLINK_MODE_GUIDED)
+        return;
+
+    if (GT_MARGIN(curr.altitude, DEST_ALT)) {
+        printf(">>>> Going back home.\n");
```
Comment (bdilly) [inline src/samples/mavlink/goto.c]: it doesn't seem to be going back home.
I've tested it here:

>>>> Taking off.
lat: -35.363262, long: 149.165237, alt: 0.090000
[...]
lat: -35.363262, long: 149.165237, alt: 9.760000
>>>> Successful takeoff, starting a new mission, heading to: (-35.361355, 149.166214, 20.000000)
[...]
lat: -35.361382, long: 149.166199, alt: 19.120001
>>>> Going back home.
lat: -35.361359, long: 149.166214, alt: 19.059999
[...]
lat: -35.361351, long: 149.166214, alt: 0.090000
>>>> Landed...

Shouldn't it go back to lat: -35.363262, long: 149.165237 ?

**Comment by bdilly** [comment]
Adding a README.md with information about samples and instructions to use them could be nice as well, just like src/samples/flow/oic/README.md

**Comment by dorileo** [comment]
ok guys, applied. Thanks.

**Pull request closed without merging**